File: tests/scanner/fixtures/python/01-basic/basic.py
"""basic.py — Python extractor fixture (PY-1..PY-3)."""
from __future__ import annotations

import os
import json as j
from typing import Dict, List, Optional, Protocol, Union
from .legacy import *
from ..pkg import helper as h
from dataclasses import dataclass
from abc import ABC, abstractmethod
from enum import Enum

MAX_USERS: int = 100
DEFAULT_NAME = "anon"


@dataclass
class User:
    """A user record."""

    id: int
    name: str
    email: Optional[str] = None
    aliases: List[str] = None
    tags: Dict[str, int] = None
    pipe_optional: str | None = None


class Repository(Protocol):
    """Structural typing surface for a user repository."""

    def find_all(self) -> List[User]: ...


class UserRepository(Repository):
    """Concrete user repository."""

    users: List[User]
    cache: Dict[int, User]
    maybe: Optional[User] = None

    def __init__(self, seed: List[User]) -> None:
        self.users = seed
        self.cache = {}
        self.private_counter = 0

    @classmethod
    def empty(cls) -> "UserRepository":
        return cls([])

    @staticmethod
    def version() -> str:
        return "1.0"

    @property
    def size(self) -> int:
        return len(self.users)

    async def fetch(self, id: int) -> Optional[User]:
        return self.cache.get(id)


class AbstractGateway(ABC):
    @abstractmethod
    def send(self, payload: Dict[str, str]) -> None: ...


class Status(Enum):
    ACTIVE = "active"
    INACTIVE = "inactive"


def top_level_helper(x: int) -> int:
    return x + 1


async def async_helper() -> None:
    pass

File: tests/scanner/fixtures/kotlin/01-basic/basic.kt
package com.example.app

import kotlin.collections.List
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.StateFlow
import java.util.Optional
import com.example.legacy.*

@Serializable
data class User(val id: Long, val name: String, val email: String?) {
    companion object {
        const val MAX_NAME = 100
    }
}

sealed interface Event {
    data class Click(val x: Int, val y: Int) : Event
    object Closed : Event
}

interface Repository<T> {
    suspend fun findAll(): List<T>
    fun observe(): Flow<T>
}

@Component
class UserRepository(
    private val users: MutableList<User>,
    private val cache: Map<Long, User>,
) : Repository<User> {
    private val state: StateFlow<User>? = null
    private val maybe: Optional<User> = Optional.empty()

    override suspend fun findAll(): List<User> = users.toList()
    override fun observe(): Flow<User> = throw NotImplementedError()
}

fun String.titleCase(): String = this.replaceFirstChar { it.uppercase() }

fun topLevelHelper(n: Int): Int = n + 1

enum class Status { ACTIVE, INACTIVE }

annotation class Audit(val level: Int = 1)

object AppConfig {
    const val VERSION = "1.0.0"
    val users: List<User> = emptyList()
}
